### PR TITLE
GIX-1187: List neurons per page

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Reduce number of update calls.

Do not load neurons unless they will be used. To accomplish that, we load the neurons in the pages they are needed.

# Changes

* Remove `listNeurons` from `app.services.ts`.
* Add `listNeurons` call in NnsNeurons.svelte.
* Add `listNeurons` call in NnsProposals.svelte if user is signed in.
* Add `listNeurons` call in NnsProposalDetail.svelte if user is signed in.

Minor change:
* `listNeurons` relied on the order of the responses to set the neurons. I changed that.

# Tests

* Test new functionality in the page tests.
* Mock import in the routes tests.
* Test new functionality in the workflows NnsProposalDetail and NnsProposals.
* Test change in `listNeurons` service.

### CODE REVIEW

**Comment by lmuntaner** [comment]
@peterpeterparker @dskloetd please review

File: frontend/src/lib/pages/NnsNeurons.svelte [hunk 1]
```diff
@@ -10,12 +10,16 @@
...
   import EmptyMessage from "$lib/components/ui/EmptyMessage.svelte";
-
-  // Neurons are fetch on page load. No need to do it in the route.
+  import { onMount } from "svelte";
+  import { listNeurons } from "$lib/services/neurons.services";
 
   let isLoading = false;
   $: isLoading = $neuronsStore.neurons === undefined;
 
+  onMount(() => {
```
Comment (dskloetd) [inline frontend/src/lib/pages/NnsNeurons.svelte:19]: Why do you check isSignedIn in the other files but not here?
And why onMount() here but not in the other files?

File: frontend/src/lib/services/neurons.services.ts [hunk 2]
```diff
@@ -243,11 +243,20 @@ export const listNeurons = async ({
   callback?: (certified: boolean) => void;
   strategy?: QueryAndUpdateStrategy;
 } = {}): Promise<void> => {
+  let certifiedNeuronsSet = false;
   return queryAndUpdate<NeuronInfo[], unknown>({
     strategy,
     request: ({ certified, identity }) =>
       neuronsApiService.queryNeurons({ certified, identity }),
     onLoad: async ({ response: neurons, certified }) => {
+      if (!certified && certifiedNeuronsSet) {
```
Comment (dskloetd) [inline frontend/src/lib/services/neurons.services.ts:252]: Aren't you duplicating logic from the api-service here?
The API service will use the cache (if present) for non-certified requests so this code here is redundant, right?

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts [hunk 3]
```diff
@@ -2,78 +2,61 @@
...
-        .spyOn(neuronsStore, "subscribe")
-        .mockImplementation(
-          buildMockNeuronsStoreSubscribe([
-            mockNeuron,
-            mockNeuron2,
-            spawningNeuron,
-          ])
-        );
+        .spyOn(authServices, "getAuthenticatedIdentity")
+        .mockImplementation(() => Promise.resolve(mockGetIdentity()));
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:39]: Do you need a new mockGetIdentity() each time? Or could you use mockResolvedValue?
Also instead of `() => Promise.resolve(mockGetIdentity())` you could use `async () => mockGetIdentity()` if you think it's nicer.

Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:39]: Why does the test need mockGetIdentity() now but it didn't need it before?

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts [hunk 4]
```diff
@@ -84,16 +67,69 @@ describe("NnsNeurons", () => {
   });
 
   describe("no neurons", () => {
-    beforeAll(() => {
+    beforeEach(() => {
+      jest.clearAllMocks();
+      neuronsStore.reset();
+      resetNeuronsApiService();
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:73]: Put all the shared setup in a separate `beforeEach` at the higher level.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts [hunk 5]
```diff
@@ -2,78 +2,61 @@
...
-
-    it("should render spawning neurons as disabled", () => {
+    it("should render spawning neurons as disabled", async () => {
       const { queryAllByTestId } = render(NnsNeurons);
 
+      // Wait for the neurons to be loaded and rendered
+      await waitFor(() => {
+        const neuronCards = queryAllByTestId("neuron-card");
+        return expect(neuronCards.length).toBe(3);
+      });
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:52]: Why did you need this now and now before?

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts [hunk 6]
```diff
@@ -2,78 +2,61 @@
...
       const disabledCards = neuronCards.filter(
         (card) => card.getAttribute("aria-disabled") === "true"
       );
       expect(disabledCards.length).toBe(1);
     });
 
-    it("should subscribe to store", () => {
-      render(NnsNeurons);
-      expect(authStoreMock).toHaveBeenCalled();
-    });
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:75]: Why remove this?

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts [hunk 7]
```diff
@@ -84,16 +67,69 @@ describe("NnsNeurons", () => {
...
+        .spyOn(authServices, "getAuthenticatedIdentity")
+        .mockImplementation(() => Promise.resolve(mockGetIdentity()));
+      jest.spyOn(api, "queryNeurons").mockResolvedValue([]);
+    });
+
+    it("should call query neurons twice when rendered", async () => {
+      resetNeuronsApiService();
+      render(NnsNeurons);
+
+      await waitFor(() =>
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:103]: I read waitFor is a bad pattern. Is there something specific we can await?

File: frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts [hunk 8]
```diff
@@ -61,25 +65,58 @@ describe("ProposalDetail", () => {
...
+        certified: false,
+      });
+    });
   });
 
-  it("should render proposal detail if signed in", async () => {
-    authStoreMock.next({
-      identity: mockIdentity,
+  describe("logged out user", () => {
+    beforeEach(() => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts:100]: Put shared setup on the higher level?

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts [hunk 9]
```diff
@@ -63,6 +67,35 @@ describe("NnsProposals", () => {
 
     afterAll(() => jest.clearAllMocks());
 
+    describe("neurons", () => {
+      beforeEach(() => {
+        // TODO: Mock the canister call instead of the service
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposals.spec.ts:72]: What is this referring to? Where is it mocking the service?

File: frontend/src/tests/lib/services/neurons.services.spec.ts [hunk 10]
```diff
@@ -319,6 +316,15 @@ describe("neurons-services", () => {
   });
 
   describe("list neurons", () => {
+    const spyQueryNeurons = jest
+      .spyOn(api, "queryNeurons")
+      .mockImplementation(() => Promise.resolve(neurons));
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/neurons.services.spec.ts:321]: mockResolvedValue?

File: frontend/src/tests/lib/services/neurons.services.spec.ts [hunk 11]
```diff
@@ -319,6 +316,15 @@ describe("neurons-services", () => {
...
+
+    beforeEach(() => {
+      resetNeuronsApiService();
+      jest.clearAllMocks();
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/neurons.services.spec.ts:325]: Put these at the highest level?

File: frontend/src/lib/pages/NnsNeurons.svelte (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/pages/NnsNeurons.svelte:19]: `NnsNeurons` is rendered only when the user is logged in.

While proposals are also rendered when user is not logged in.

In the others, I want to call list neurons if the authStore changes, that's why I don't put it in the onMount.

Here, it's enough to load it when the component is rendered.

File: frontend/src/lib/services/neurons.services.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/lib/services/neurons.services.ts:252]: Not exactly. The store does not separate data with certified and not certified.

```
{
  neurons: NeuronInfo[],
  certified: boolean,
}
```

Therefore, if the query was coming after the update, the `neuronStore.setData` was setting `certified` to `false` and changing the `neurons` for the uncertified ones.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:39]: Same thing

Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:39]: I was mocking `listNeurons`.

Now I'm mocking the api layer.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:73]: No, I want it after each of the `it` tests also.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 5 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:52]: Because I was filling the store and there was no need to wait for this. The first render already came with the cards.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:75]: I don't understand why it was here, this is not done at this level.

Maybe it was happening in some child component and therefore it passed the test. But the NnsNeurons does not subscribe to the store.

File: frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts (hunk 8 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts:100]: I needed it being reset after each `it`

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 9 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:72]: Before, it was mocking the governanceCanister, but now we moved to a pattern to mock the api layer.

It's a little bit inconsistent like this. But I did't want to refactor much more.

File: frontend/src/tests/lib/services/neurons.services.spec.ts (hunk 10 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/services/neurons.services.spec.ts:321]: yes

File: frontend/src/tests/lib/services/neurons.services.spec.ts (hunk 11 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/services/neurons.services.spec.ts:325]: I needed them reset after each `it`

**Comment by lmuntaner** [comment]
@dskloetd ready for another review

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 7 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:103]: Sorry, maybe what I read was about `tick`. I thought it was in the Svelte tutorial but I can't find it now. Nevermind.

File: frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts (hunk 8 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts:100]: After or before?
Either way, it will do that on the higher level. Just for more `it`s rather than only the ones in this block.

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 9 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:72]: I still don't understand. Is this comment referring to something in a different file? Or is there a specific line in this file where it's mocking the service and it shouldn't?

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts [hunk 12]
```diff
@@ -63,6 +67,35 @@ describe("NnsProposals", () => {
...
+        const mockGovernanceCanister: MockGovernanceCanister =
+          new MockGovernanceCanister([]);
+        jest
+          .spyOn(GovernanceCanister, "create")
+          .mockImplementation((): GovernanceCanister => mockGovernanceCanister);
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposals.spec.ts:77]: mockResolvedValue?

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts [hunk 13]
```diff
@@ -195,6 +228,28 @@ describe("NnsProposals", () => {
...
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposals.spec.ts:238]: mockResolvedValue?

File: frontend/src/lib/services/neurons.services.ts (hunk 2 above)
Comment (dskloetd) [reply frontend/src/lib/services/neurons.services.ts:252]: Yes, it will respond to both from the cache. But it shouldn't be a problem that you get 2 responses both from the cache, right?

Comment (lmuntaner) [reply frontend/src/lib/services/neurons.services.ts:252]: Not in this test, I reset the cache here.

I found the problem, but I couldn't replicate it in isolation.

I added some comment and a todo. I also did the test in the bad case that query comes after update call and it still works.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:73]: Changed!

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:39]: forget what I said. It can be greatly simplified:

```
      jest
        .spyOn(authServices, "getAuthenticatedIdentity")
        .mockResolvedValue(mockIdentity);
```

File: frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts (hunk 8 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts:100]: Same as before, I thought that if you put it out of the `description` the beforeEach applied to each description

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 9 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:72]: Below we do this:

```
const mockGovernanceCanister: MockGovernanceCanister =
          new MockGovernanceCanister([]);
jest
          .spyOn(GovernanceCanister, "create")
          .mockImplementation((): GovernanceCanister => mockGovernanceCanister);
```

Instead of mocking directly mocking the api:

```
jest.spyOn(proposalsApi, "queryProposals")...
```

I didn't want to change how it's done in this PR.

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 12 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:77]: yes

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 13 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts]: done

**Comment by lmuntaner** [comment]
@dskloetd ready for another review

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 4 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:74]: You can move `resetNeuronsApiService();` as well.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts [hunk 14]
```diff
@@ -84,16 +70,65 @@ describe("NnsNeurons", () => {
...
+        })
+      );
+      expect(api.queryNeurons).toHaveBeenCalledWith({
+        identity: mockIdentity,
+        certified: false,
+      });
+
+      render(NnsNeurons);
+
+      await waitFor(() => expect(api.queryNeurons).toHaveBeenCalledTimes(2));
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:131]: How doe this guarantee that it's not called more *after* this?

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts [hunk 15]
```diff
@@ -63,6 +67,35 @@ describe("NnsProposals", () => {
...
 
+    describe("neurons", () => {
+      beforeEach(() => {
+        // TODO: Mock the canister call instead of the service
+        const mockGovernanceCanister: MockGovernanceCanister =
+          new MockGovernanceCanister([]);
+        jest
+          .spyOn(GovernanceCanister, "create")
+          .mockReturnValue(mockGovernanceCanister);
+        jest.clearAllMocks();
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposals.spec.ts:78]: It's safer to use resetAllMocks because clearAllMock doesn't reset the defined behavior. Though you'd have to do it earlier or you would reset the behavior you just defined on the previous lines.
But I would anyway suggest to put all the resets at the top level beforeEach (line 51).

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts [hunk 16]
```diff
@@ -195,6 +228,28 @@ describe("NnsProposals", () => {
...
+        neuronsStore.reset();
+        resetNeuronsApiService();
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposals.spec.ts:241]: You don't need the resets here once you put them all on the top level.

File: frontend/src/tests/lib/services/neurons.services.spec.ts [hunk 17]
```diff
@@ -319,6 +316,14 @@ describe("neurons-services", () => {
   });
 
   describe("list neurons", () => {
+    const spyQueryNeurons = jest
+      .spyOn(api, "queryNeurons")
+      .mockResolvedValue(neurons);
+
+    beforeEach(() => {
+      resetNeuronsApiService();
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/neurons.services.spec.ts:324]: Put it on line 219 instead?

File: frontend/src/tests/lib/services/neurons.services.spec.ts [hunk 18]
```diff
@@ -331,6 +336,24 @@ describe("neurons-services", () => {
...
+      });
+
+      expect(spyQueryNeurons).toHaveBeenCalledWith({
+        identity: mockIdentity,
+        certified: false,
+      });
+
+      await listNeurons();
+
+      expect(spyQueryNeurons).toHaveBeenCalledTimes(2);
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/neurons.services.spec.ts:354]: Put a copy of this line also before line 352 to make it clear that the call on line 352 didn't change anything.

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts [hunk 19]
```diff
@@ -55,23 +59,46 @@ describe("Proposal detail page when not logged in user", () => {
 
   describe("when logged in user", () => {
     beforeEach(() => {
+      jest.clearAllMocks();
```
Comment (dskloetd) [inline frontend/src/tests/workflows/NnsProposalDetail.spec.ts:62]: Use resetAllMock instead?

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 9 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:72]: But I don't see a service?

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts [hunk 20]
```diff
@@ -55,23 +59,46 @@ describe("Proposal detail page when not logged in user", () => {
...
+        .spyOn(governanceApi, "queryNeurons")
+        .mockImplementation(async ({ certified }) => {
+          // Mock delay in one of both calls.
+          // Otherwise there is a race condition that the onLoad of queryAndUpdate for query after certified has been called.
+          // The problem seems to be that the `certifiedDone` is set in the finally.
+          // Yet, I couldn't replicate the issue with a test in `utils.services.spec.ts`.
+          // Therefore, I delay here the QUERY to prove that it still works, even in the bad case that certified data comes before uncertified.
+          if (!certified) {
+            return new Promise((resolve) =>
+              setTimeout(() => {
```
Comment (dskloetd) [inline frontend/src/tests/workflows/NnsProposalDetail.spec.ts:75]: It would be nicer if you return unresolved promises and explicitly resolve one before the other.

File: frontend/src/tests/workflows/NnsProposals.spec.ts [hunk 21]
```diff
@@ -30,19 +33,24 @@ jest.mock("$lib/api/proposals.api", () => {
...
+
 describe("NnsProposals", () => {
   afterEach(() => {
     jest.clearAllMocks();
   });
 
   describe("when signed in user", () => {
     beforeEach(() => {
+      neuronsStore.reset();
+      resetNeuronsApiService();
```
Comment (dskloetd) [inline frontend/src/tests/workflows/NnsProposals.spec.ts:46]: Put the resets at the higher level?

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 14 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:131]: I'm not sure how to test that.

Shall I add some `await tick()` and then do this test?

Comment (dskloetd) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:131]: Is there something you can do in the component (click around like the user using the component) after which you check the number of calls?

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 9 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:72]: discussed in slack, comment will change

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 16 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:241]: yes, I'll move in the top most `beforeEach`

File: frontend/src/tests/lib/services/neurons.services.spec.ts (hunk 17 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/services/neurons.services.spec.ts:324]: Ok

File: frontend/src/tests/lib/services/neurons.services.spec.ts (hunk 18 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/services/neurons.services.spec.ts:354]: ok

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 19 above)
Comment (lmuntaner) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts:62]: As mentioned, there are some clash with mocks added in a global file. I'll keep it and that can be improved later

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 20 above)
Comment (lmuntaner) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts:75]: Not sure I understand.

Here I'm resolving one later than the other.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 14 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:137]: Not in this list part. I added some `await tick();`s. This "simulates" the user waiting...

File: frontend/src/tests/workflows/NnsProposals.spec.ts (hunk 21 above)
Comment (lmuntaner) [reply frontend/src/tests/workflows/NnsProposals.spec.ts:46]: done

**Comment by lmuntaner** [comment]
@dskloetd ready for another review

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts [hunk 22]
```diff
@@ -2,78 +2,65 @@
...
-    authStoreMock = jest
-      .spyOn(authStore, "subscribe")
-      .mockImplementation(mockAuthStoreSubscribe);
+    jest.resetAllMocks();
+    neuronsStore.reset();
   });
 
   describe("with enough neurons", () => {
     beforeEach(() => {
+      resetNeuronsApiService();
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:27]: Put this on line 23?

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 4 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:75]: In the diff I see it's still here.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts [hunk 23]
```diff
@@ -84,16 +71,70 @@ describe("NnsNeurons", () => {
...
+        })
+      );
+      expect(api.queryNeurons).toHaveBeenCalledWith({
+        identity: mockIdentity,
+        certified: false,
+      });
+    });
+
+    it("should NOT call query neurons after being visited", async () => {
+      resetNeuronsApiService();
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsNeurons.spec.ts:116]: If you have this in beforeEach, you shouldn't need this here, right?

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 9 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:67]: It still looks the same to me?

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts [hunk 24]
```diff
@@ -191,10 +216,29 @@ describe("NnsProposals", () => {
...
           return () => undefined;
         });
+      jest.spyOn(governanceApi, "queryNeurons").mockResolvedValue([]);
     });
 
     afterAll(() => jest.clearAllMocks());
 
+    describe("neurons", () => {
+      beforeEach(() => {
+        // TODO: Mock the canister call instead of the service
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposals.spec.ts:226]: same comment here

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 19 above)
Comment (dskloetd) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts:62]: OK, but you don't need it here because you have it on line 56 already, right?

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts [hunk 25]
```diff
@@ -55,23 +59,46 @@ describe("Proposal detail page when not logged in user", () => {
 
   describe("when logged in user", () => {
     beforeEach(() => {
+      jest.clearAllMocks();
+      neuronsStore.reset();
```
Comment (dskloetd) [inline frontend/src/tests/workflows/NnsProposalDetail.spec.ts:63]: You have this on line 57 already?

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts [hunk 26]
```diff
@@ -55,23 +59,46 @@ describe("Proposal detail page when not logged in user", () => {
...
+      resetNeuronsApiService();
```
Comment (dskloetd) [inline frontend/src/tests/workflows/NnsProposalDetail.spec.ts:64]: Move this to line 58?

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 20 above)
Comment (dskloetd) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts:75]: Make the promises available outside the implementation in the test and explicitly resolve them there in a specific order.

So like this:
```
let resolveCertifiedPromise;
let resolveUncertifiedPromise;

      jest
        .spyOn(governanceApi, "queryNeurons")
        .mockImplementation(async ({ certified }) => {
          return new Promise((resolve) => {
            if (certified) {
             resolveCertifiedPromise = resolve;
            } else {
              resolveUncertifiedPromise = resolve;
            }
          });
       });
```

In the test:
```
...
resolveCertifiedPromise();
...
resolveUncertifiedPromise();
...
```

Does this make sense?

File: frontend/src/tests/workflows/NnsProposals.spec.ts (hunk 21 above)
Comment (dskloetd) [reply frontend/src/tests/workflows/NnsProposals.spec.ts:46]: I still see it here.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 23 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts:116]: correct, moving it

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 9 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts:67]: I reverted some changes because test was failing and I forgot to change it back, sorry.

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 20 above)
Comment (lmuntaner) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts:75]: aahh, yes, I understand now.

File: frontend/src/tests/lib/pages/NnsNeurons.spec.ts (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsNeurons.spec.ts]: done

File: frontend/src/tests/lib/pages/NnsProposals.spec.ts (hunk 9 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/NnsProposals.spec.ts]: done

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 19 above)
Comment (lmuntaner) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts]: done

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 20 above)
Comment (lmuntaner) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts]: done

**Comment by lmuntaner** [comment]
@dskloetd please review

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts [hunk 27]
```diff
@@ -47,31 +52,69 @@ jest.mock("$lib/utils/html.utils", () => ({
...
+              resolveUncertifiedPromise = resolve;
+            }
+          });
+        });
       jest
         .spyOn(authStore, "subscribe")
         .mockImplementation(mockAuthStoreSubscribe);
     });
 
     it("should render proposal with certified data", async () => {
```
Comment (dskloetd) [inline frontend/src/tests/workflows/NnsProposalDetail.spec.ts:90]: Is the purpose of this test to test what happens if the promises resolve in the "wrong" order? Or are you piggybacking this on another test?

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts [hunk 28]
```diff
@@ -94,10 +139,40 @@ describe("Proposal detail page when not logged in user", () => {
...
+      });
+      // Make sure we have the resolve functions.
+      await waitFor(() =>
+        expect(typeof resolveCertifiedPromise).toBe("function")
+      );
+      await waitFor(() =>
+        expect(typeof resolveUncertifiedPromise).toBe("function")
+      );
+      resolveUncertifiedPromise([mockNeuron]);
+      resolveCertifiedPromise([mockNeuron]);
```
Comment (dskloetd) [inline frontend/src/tests/workflows/NnsProposalDetail.spec.ts:158]: Does this test also rely on specific control over the order of resolution?
Otherwise would it make sense to use that specific behavior only for that test that needs it?

**Review by dskloetd — APPROVED** [review]
I don't want to block on the remaining comments but they do point out a real potential issue.

If there is a specific issue that needs to be tested, it needs its own separate test, otherwise it's unclear and may be removed later.

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 28 above)
Comment (lmuntaner) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts:158]: I'm not trying to test that here.

File: frontend/src/tests/workflows/NnsProposalDetail.spec.ts (hunk 27 above)
Comment (lmuntaner) [reply frontend/src/tests/workflows/NnsProposalDetail.spec.ts:90]: No, that's not the purpose of this test.

I found a test for that in `utils.services.spec.ts` already
